lib/grandpa/message_handler.go

```diff
@@ -280,21 +303,50 @@ func (h *MessageHandler) verifyCommitMessageJustification(fm *CommitMessage) err
 }
 
 func (h *MessageHandler) verifyPreVoteJustification(msg *CatchUpResponse) (common.Hash, error) {
+	voters := make(map[ed25519.PublicKeyBytes]map[common.Hash]int, len(msg.PreVoteJustification))
+	eqVotesByHash := make(map[common.Hash]map[ed25519.PublicKeyBytes]struct{})
+
+	// identify equivocatory votes by hash
+	for _, justification := range msg.PreVoteJustification {
+		if _, ok := voters[justification.AuthorityID]; !ok {
+			voters[justification.AuthorityID] = make(map[common.Hash]int)
+		}
+
+		voters[justification.AuthorityID][justification.Vote.Hash]++
+
+		if voters[justification.AuthorityID][justification.Vote.Hash] > 1 {
+			if _, ok := eqVotesByHash[justification.Vote.Hash]; !ok {
+				eqVotesByHash[justification.Vote.Hash] = make(map[ed25519.PublicKeyBytes]struct{})
+			}
+
+			eqVotesByHash[justification.Vote.Hash][justification.AuthorityID] = struct{}{}
+		}
+	}
+
 	// verify pre-vote justification, returning the pre-voted block if there is one
```

Comment:
I have a feeling this function is actually incorrect (not because of you, but it was incorrect before), as it doesn't calculate the grandpa-GHOST by checking for vote ancestor blocks, it just directly checks if a block had >2/3 votes. don't worry about this for now, unless you want to fix it haha. it should be fixed by #1531

Reply:
Got it!

Reply:
@noot this pr is blocked by #1531 or it can be merged?

Reply:
it can be merged!